crates/bitcoind_rpc/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+pub use bitcoincore_rpc;
+use bitcoincore_rpc::bitcoincore_rpc_json;
+
+/// Represents a transaction that exists in the mempool.
+pub struct MempoolTx {
+    /// The transaction.
+    pub tx: Transaction,
+    /// Time when transaction first entered the mempool (in epoch seconds).
+    pub time: u64,
+}
+
+/// A block obtained from `bitcoind`.
+pub struct EmittedBlock {
+    /// The actual block.
+    pub block: Block,
+    /// The height of the block.
+    pub height: u32,
+}
+
+/// A block header obtained from `bitcoind`.
+pub struct EmittedHeader {
+    /// The actual block header.
+    pub header: Header,
+    /// The height of the block header.
+    pub height: u32,
+}
+
+/// A structure that emits data sourced from [`bitcoincore_rpc::Client`].
+///
+/// Refer to [module-level documentation] for more.
+///
+/// [module-level documentation]: crate
+pub struct Emitter<'c, C> {
+    client: &'c C,
+    start_height: u32,
+
+    emitted: BTreeMap<u32, BlockHash>,
+    last_block: Option<bitcoincore_rpc_json::GetBlockResult>,
+    /// Records the latest first-seen epoch of emitted mempool transactions.
+    ///
+    /// This allows us to avoid re-emitting some mempool transactions. Mempool transactions need to
+    /// be re-emitted if the latest block that may contain the transaction's ancestors have not yet
+    /// been emitted.
+    last_mempool_time: usize,
+}
+
+impl<'c, C: bitcoincore_rpc::RpcApi> Emitter<'c, C> {
+    /// Constructs a new [`Emitter`] with the provided [`bitcoincore_rpc::Client`].
+    ///
+    /// `start_height` is the block height to start emitting blocks from.
+    pub fn new(client: &'c C, start_height: u32) -> Self {
+        Self {
+            client,
+            start_height,
+            emitted: BTreeMap::new(),
+            last_block: None,
+            last_mempool_time: 0,
+        }
+    }
+
+    /// Emit mempool transactions.
+    ///
+    /// This avoids re-emitting transactions (where viable). We can do this if all blocks
+    /// containing ancestor transactions are already emitted.
+    pub fn mempool(&mut self) -> Result<Vec<MempoolTx>, bitcoincore_rpc::Error> {
+        let client = self.client;
+        let prev_block_height = self.last_block.as_ref().map_or(0, |r| r.height);
+        let prev_mempool_time = self.last_mempool_time;
+        let mut latest_time = self.last_mempool_time;
+
+        let txs_to_emit = client
+            .get_raw_mempool_verbose()?
+            .into_iter()
+            .filter_map(
+                move |(txid, tx_entry)| -> Option<Result<_, bitcoincore_rpc::Error>> {
+                    let tx_time = tx_entry.time as usize;
+                    if tx_time > latest_time {
+                        latest_time = tx_time;
+                    }
+
+                    // Avoid emitting transactions that are already emitted if we can guarantee
+                    // blocks containing ancestors are already emitted. We only check block height
```

Comment:
This comment doesn't make sense to me and I don't see anything where you guarantee blocks containing ancestors are already emitted (but also don't understand why that's relevant).

Reply:
The `Indexer` may think a transaction is irrelevant for the time being (since it does not have the prev outpoints that it owns). We should keep re-emitting such a transaction.

A bitcoind's tx entry gives us the block height in which the tx is introduced to the mempool. If our `Emitter`'s last-emitted-block-state meets or surpasses this block height, we can guarantee that we have scanned or ancestors (except in case of reorg).

For reorg, we force-re-emit everything.